### PR TITLE
notifications for direct messages

### PR DESCRIPTION
**Pull request by hugegulp**
Addresses feature request: 
[Trigger notifications on direct message](https://github.com/squidowl/halloy/issues/491)

### CODE REVIEW

File: data/src/client.rs [hunk 1]
```diff
@@ -636,6 +637,16 @@ impl Client {
                             // If we sent (echo) & context exists (we sent from this client), ignore
                             return None;
                         }
+
+                        // send direct message notification
+                        // TODO: check if the active buffer is the same as who sent the message
```
Comment (qauff) [inline data/src/client.rs:642]: @tarkah @casperstorm @andymandias is there existing code/example that checks/references the active/focused buffer's information?

Comment (casperstorm) [reply data/src/client.rs:642]: Take a look here @hugegulp: https://github.com/squidowl/halloy/blob/dbac8aeb4910095760d93ef19eab109eb3d7864c/src/screen/dashboard/sidebar.rs#L281-L290

`panes` and `focus` is in the `Dashboard` struct. I think what you could do is make a helper `fn` on the `Dashboard` struct which returns the current selected `Pane` (essentially refactoring that piece of code i linked into a `fn`). You can then check the `buffer` of that pane to see if its the one you got a message for.

Once you have such a method you can call it on the `dashboard` which you can get like this:
https://github.com/squidowl/halloy/blob/f0a4e8e4baeea72dfbe1c48f93e29ba1eb03dc70/src/main.rs#L408-L410

Comment (casperstorm) [reply data/src/client.rs:642]: Alternative, and potentially better, take a look at:

https://github.com/squidowl/halloy/blob/10fde389a92edf4b1f29e2a132abbb771fc88fa0/data/src/history/manager.rs#L250-L265

Comment (hugegulp) [reply data/src/client.rs]: This is a great suggestion. I went with the history manager method! I added the conditional check in `src/main.rs` let me know what you think!

**Comment by hugegulp** [comment]
The only scenario I haven't covered for receiving a notification sound is when halloy itself is not in focus. I've noticed when I have a DM/query open and I navigate away from halloy, when I get a message, the notification doesn't play. So next stage of this feature would be to figure out if halloy is in focus. Is that possible currently, @casperstorm?

**Comment by hugegulp** [comment]
> So next stage of this feature would be to figure out if halloy is in focus.

I guess another way to look at this is, to make it so the history manager counts messages as unread if halloy (and panes) are not in focus.

**Comment by casperstorm** [comment]
https://github.com/iced-rs/iced/blob/master/core/src/window/event.rs#L39-L43

We can detect if a window has focus with these events.

**Comment by hugegulp** [comment]
> https://github.com/iced-rs/iced/blob/master/core/src/window/event.rs#L39-L43
> 
> We can detect if a window has focus with these events.

Thanks for this @casperstorm I've pushed up an implementation of tracking the window's focus, and triggering notifications for DMs if the window is out of focus! We should be good to go now! :pray:

**Comment by hugegulp** [comment]
Apologies for all the little refactors! _Now_ I'm ready for review :pray:

**Review by casperstorm — APPROVED** [review]
This is really well done, @tarkah you have any feedback before merge?

**Review by tarkah — APPROVED** [review]
Great addition, thanks! Added a couple small changes.

**Comment by tarkah** [comment]
Fuck I messed up pushing my changes... will re-open

**Pull request closed without merging**